subsys/logging/log_backend_rpc.c

```diff
@@ -0,0 +1,310 @@
+ * 1. log added
+ * 2. log sent over nRF RPC
+ * 3. more logs generated by nRF RPC
+ * 4. more logs sent over nRF RPC
+ * ...
+ */
+static const char *const filtered_out_sources[] = {
+	"nrf_rpc",
+	"NRF_RPC",
+};
+
+static bool panic_mode;
+static uint32_t log_format = LOG_OUTPUT_TEXT;
+static uint8_t current_level;
+static uint8_t log_output_buffer[CONFIG_LOG_BACKEND_RPC_BUFFER_SIZE];
+
+#ifdef CONFIG_LOG_BACKEND_RPC_CRASH_LOG
+/* Header for the crash log data stored in the retained RAM partition. */
+typedef struct crash_log_header {
+	size_t size;
+} crash_log_header_t;
+
+static size_t crash_log_cur_size;
+static size_t crash_log_max_size;
+
+static void crash_log_clear(void)
+{
+	const struct device *const crash_log = CRASH_LOG_DEV;
+	ssize_t rc;
+
+	crash_log_cur_size = 0;
+	crash_log_max_size = 0;
+
+	rc = retention_clear(crash_log);
+
+	if (rc) {
+		return;
+	}
+
+	rc = retention_size(crash_log);
+
+	if (rc <= (ssize_t)sizeof(crash_log_header_t)) {
+		/* Either failed to obtain the retained partition size, or it is too small. */
+		return;
+	}
+
+	crash_log_max_size = (size_t)rc - sizeof(crash_log_header_t);
+}
+
+static void retain(const uint8_t *data, size_t length)
+{
+	const struct device *const crash_log = CRASH_LOG_DEV;
+	crash_log_header_t header;
+	int rc;
+
+	if (crash_log_cur_size >= crash_log_max_size) {
+		/* Crash log partition overflow. */
+		return;
+	}
+
+	length = MIN(length, crash_log_max_size - crash_log_cur_size);
+
+	rc = retention_write(crash_log, sizeof(header) + crash_log_cur_size, data, length);
+
+	if (rc) {
+		return;
+	}
+
+	crash_log_cur_size += length;
+	header.size = crash_log_cur_size;
+
+	(void)retention_write(crash_log, 0, (const uint8_t *)&header, sizeof(header));
+}
+
+static void log_rpc_get_crash_log_handler(const struct nrf_rpc_group *group,
+					  struct nrf_rpc_cbor_ctx *ctx, void *handler_data)
+{
+	const struct device *const crash_log = CRASH_LOG_DEV;
+	size_t offset;
+	size_t length;
+	bool decoded_ok;
+	crash_log_header_t header;
+	struct nrf_rpc_cbor_ctx rsp_ctx;
+
+	decoded_ok = zcbor_uint_decode(ctx->zs, &offset, sizeof(offset));
+	decoded_ok = decoded_ok && zcbor_uint_decode(ctx->zs, &length, sizeof(length));
+	nrf_rpc_cbor_decoding_done(group, ctx);
+
+	if (!decoded_ok) {
+		goto err;
+	}
+
+	if (retention_is_valid(crash_log) != 1) {
+		goto err;
+	}
+
+	if (retention_read(crash_log, 0, (uint8_t *)&header, sizeof(header))) {
+		goto err;
+	}
+
+	if (offset >= header.size) {
+		goto err;
+	}
+
+	length = MIN(length, header.size - offset);
+	NRF_RPC_CBOR_ALLOC(group, rsp_ctx, 5 + length);
+
+	if (!zcbor_bstr_start_encode(rsp_ctx.zs)) {
+		goto discard;
+	}
+
+	if (retention_read(crash_log, sizeof(header) + offset, rsp_ctx.zs[0].payload_mut, length)) {
+		goto discard;
+	}
+
+	rsp_ctx.zs[0].payload_mut += length;
+
+	if (!zcbor_bstr_end_encode(rsp_ctx.zs, NULL)) {
+		goto discard;
+	}
+
+	goto send;
+
+discard:
+	NRF_RPC_CBOR_DISCARD(group, rsp_ctx);
+err:
+	/* Send empty string */
```

Comment:
Why is the empty string required?

Reply:
Because the API contract which I documented is that an empty string means "end of file". We could return an error code as well but for a caller it doesn't really make a difference why the crash log is not accessible (I thought).